src/VCR/MatchObserver.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace VCR;
+
+class MatchObserver
+{
+    // This is an array keyed by a hash of the (first aka stored) request.
+    private $requestToMatchersMap;
+    private $requestCount0;
```

Comment:
I'm a bit confused with all the variables, `requestToRequestsMap`, `requestCountMap`, `requestToMatcherMap`.

`shouldObserver`: I wonder if we should provide a way to disable it. After all we are normally in a 'testing' mode and it would be very useful to always have that information. Performance I guess shouldn't be a big problem since it is not meant for production and only happens when a request does not match.